crates/bin/prove_block/src/state_utils.rs

```diff
@@ -42,52 +38,68 @@ pub(crate) async fn get_processed_state_update(
-    let accessed_addresses: HashSet<Felt252> = address_to_class_hash
-        .keys()
-        .copied()
-        .chain(address_to_nonce.keys().cloned())
-        .chain(storage_updates.keys().cloned())
-        .collect();
-
-    FormattedStateUpdate {
-        address_to_class_hash,
-        address_to_nonce,
-        class_hash_to_compiled_class_hash,
-        storage_updates,
-        accessed_addresses,
-    }
+    Ok(FormattedStateUpdate { class_hash_to_compiled_class_hash, compiled_classes: compiled_contract_classes })
 }
 
-/// Format StateDiff's NonceUpdate to a HashMap<contract_address, nonce>
-fn format_state_update_nonces(
-    state_diff: &StateDiff,
-    transactions: &[InternalTransaction],
-) -> HashMap<Felt252, Felt252> {
-    let address_to_nonce: HashMap<Felt252, Felt252> = state_diff
-        .nonces
-        .iter()
-        .map(|nonce_update| {
-            // derive original nonce
-            // TODO: understand what is going on here XD
-            let num_nonce_bumps =
-                Felt252::from(transactions.iter().fold(0, |acc, tx| {
-                    acc + if tx.sender_address == Some(nonce_update.contract_address) { 1 } else { 0 }
-                }));
-            assert!(nonce_update.nonce > num_nonce_bumps);
-            let previous_nonce = nonce_update.nonce - num_nonce_bumps;
-            (nonce_update.contract_address, previous_nonce)
-        })
-        .collect();
-    address_to_nonce
+/// This function processes a set of accessed contract addresses to retrieve their
+/// corresponding class hashes and compile them into `GenericCasmContractClass`.
+/// If the class is already present in `address_to_class_hash`, it is used directly;
+/// otherwise, it is fetched from the provided `JsonRpcClient`.
+///
+/// The resulting compiled classes and any associated mappings are returned, while
+/// the `class_hash_to_compiled_class_hash` map is updated with new entries.
+/// TODO: Handle deprecated classes
+async fn build_compiled_class_and_maybe_update_class_hash_to_compiled_class_hash(
+    provider: &JsonRpcClient<HttpTransport>,
+    block_id: BlockId,
+    accessed_addresses: &HashSet<Felt252>,
+    address_to_class_hash: &HashMap<Felt252, Felt252>,
+    class_hash_to_compiled_class_hash: &mut HashMap<Felt252, Felt252>,
+) -> Result<(HashMap<Felt252, GenericCasmContractClass>, HashMap<Felt252, GenericDeprecatedCompiledClass>), Error> {
+    let mut compiled_contract_classes: HashMap<Felt252, GenericCasmContractClass> = HashMap::new();
+    // TODO: Handle deprecated classes
+    let mut _deprecated_compiled_contract_classes: HashMap<Felt252, GenericDeprecatedCompiledClass> = HashMap::new();
+
+    for contract_address in accessed_addresses {
+        let class_hash = match address_to_class_hash.get(contract_address) {
+            Some(class_hash) => class_hash,
+            None => &provider.get_class_hash_at(block_id, contract_address).await?,
```

Comment:
Note that with anyhow, you can add `with_context("<some explanation>")` to any `Result` value so that you get a human-readable message + details on failure.